### PR TITLE
Update user database provisioner to wipe/recreate database on run and run database schema migrations in CI

### PR DESCRIPTION
**Pull request by Jkovarik**
**Summary:** Summary of changes

Addresses [CUMULUS-2122](https://bugs.earthdata.nasa.gov/browse/CUMULUS-2122)

## Changes

* Update db-provision-user-database to remove existing database and recreate. This will functionally cause data-persistence CI to wipe the database each run in prep to run migrations.
* Add db-migration module to CI

## PR Checklist

- [ ] Update CHANGELOG
- [ ] Unit tests
- [ ] Adhoc testing
- [ ] Integration tests

### CODE REVIEW

File: bamboo/bootstrap-tf-deployment.sh [hunk 1]
```diff
@@ -56,7 +56,31 @@ echo "Deploying Cumulus data-persistence module to $DEPLOYMENT"
...
+    region = \"$AWS_REGION\"
+    dynamodb_table = \"$TFSTATE_LOCK_TABLE\"
+  }
+}" >> ci_backend.tf
 
+# Initialize deployment
+../terraform init \
+  -input=false
+
+echo "Deploying the Cumulus data migratino module to $DEPLOYMENT"
```
Comment (markdboyd) [inline bamboo/bootstrap-tf-deployment.sh:74]: 🐜 
```suggestion
echo "Deploying the Cumulus data migration module to $DEPLOYMENT"
```

File: example/lambdas/db-provision-user-database/src/index.ts [hunk 2]
```diff
@@ -43,18 +43,23 @@ export const handler = async (event: HandlerEvent): Promise<void> => {
...
-      await knex.raw(`create database "${dbUser}_db";`);
-      await knex.raw(`grant all privileges on database "${dbUser}_db" to "${dbUser}"`);
     } else {
       await knex.raw(`alter user "${dbUser}" with encrypted password '${event.dbPassword}'`);
-      await knex.raw(`grant all privileges on database "${dbUser}_db" to "${dbUser}"`);
     }
 
+    if (dbExistsResults.length !== 0) {
+      await knex.raw(`alter database "${dbUser}_db" connection limit 0;`);
+      await knex.raw(`select pg_terminate_backend(pg_stat_activity.pid) from pg_stat_activity where pg_stat_activity.datname = '${dbUser}_db'`);
```
[the remark this answers is not in the document]
Comment (Jkovarik) [reply example/lambdas/db-provision-user-database/src/index.ts:57]: You can't drop a database with open connections - we have to stop new connections and kill the open ones.   I've ad-hoc tested this works locally and in the cluster, at least with a few connections.

File: example/lambdas/db-provision-user-database/tests/test-index.js [hunk 3]
```diff
@@ -87,29 +93,82 @@ test('provision user fails if invalid user string is used', async (t) => {
...
-  testUserKnex.destroy();
+
+  // Validate the table exists in the created database
+  const validateCreateTable = await t.context.testKnex.raw(tableExistsQuery);
+  t.is(validateCreateTable.rows[0].tablename, 'testTable');
+
+  await handler(handlerEvent);
+  t.context.testKnex.destroy();
+
+  t.context.testKnex = Knex({ ...knexConfig, password: handlerEvent.dbPassword });
```
[the remark this answers is not in the document]
Comment (markdboyd) [reply example/lambdas/db-provision-user-database/tests/test-index.js:165]: Better yet: why do we have to destroy and re-create the connection?

File: example/lambdas/db-provision-user-database/tests/test-index.js [hunk 4]
```diff
@@ -53,7 +53,13 @@ test.afterEach.always(async (t) => {
   await knex.raw(`drop user if exists "${t.context.expectedDbUser}"`);
 });
 
-test('provision user database creates the expected database', async (t) => {
+test.afterEach.always(async (t) => {
```
Comment (markdboyd) [inline example/lambdas/db-provision-user-database/tests/test-index.js:56]: There's an `afterEach.always` block already above this? I assume you can't have both

File: example/lambdas/db-provision-user-database/tests/test-index.js [hunk 5]
```diff
@@ -87,29 +93,82 @@ test('provision user fails if invalid user string is used', async (t) => {
...
+      password: 'testPassword',
+      database: expectedTestDb,
+    },
+  };
+
   await handler(handlerEvent);
+  t.context.testKnex = Knex(knexConfig);
+  let heartBeat = await t.context.testKnex.raw('SELECT 1');
+  t.is(heartBeat.rowCount, 1);
+  t.context.testKnex.destroy();
```
Comment (markdboyd) [inline example/lambdas/db-provision-user-database/tests/test-index.js:117]: Does this need to be `await`ed throughout?

File: example/lambdas/db-provision-user-database/tests/test-index.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (Jkovarik) [reply example/lambdas/db-provision-user-database/tests/test-index.js:165]: > Better yet: why do we have to destroy and re-create the connection?

The database forcibly drops the connection, which causes a knex error. Yes - this was rewritten into two tests, it originally had the password change as well.    I just missed the reconfiguration - will update.

[the remark this answers is not in the document]
Comment (Jkovarik) [reply example/lambdas/db-provision-user-database/tests/test-index.js:165]: > Has the dbPassword changed? Would t.context.testKnex = Knex(knexConfig); not work?

Yes, it was originally one test, I just missed that in the refactor, will sort.

File: example/lambdas/db-provision-user-database/tests/test-index.js (hunk 4 above)
Comment (Jkovarik) [reply example/lambdas/db-provision-user-database/tests/test-index.js:56]: That actually should be 477364b5a -- given this is a throwaway unit container, we don't want to kill the postgres state on a failed test: 

477364b5a

File: example/lambdas/db-provision-user-database/tests/test-index.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (Jkovarik) [reply example/lambdas/db-provision-user-database/tests/test-index.js]: e3ca5077d

File: example/lambdas/db-provision-user-database/tests/test-index.js (hunk 5 above)
Comment (Jkovarik) [reply example/lambdas/db-provision-user-database/tests/test-index.js:117]: That's probably good hygiene yes: fb5e0b545

File: example/lambdas/db-provision-user-database/tests/test-index.js (hunk 4 above)
Comment (Jkovarik) [reply example/lambdas/db-provision-user-database/tests/test-index.js:56]: Yes, and in fact you can have multiples of each per the docs:

https://github.com/avajs/ava/blob/master/docs/01-writing-tests.md

For peace of mind I just went and threw a few errors into the block and got the failures/counts I'd expect.   The databases are also cleaning up each time from what I've observed.

File: example/lambdas/db-provision-user-database/tests/test-index.js (hunk 5 above)
Comment (markdboyd) [reply example/lambdas/db-provision-user-database/tests/test-index.js:117]: I think this hasn't been pushed up yet

Comment (Jkovarik) [reply example/lambdas/db-provision-user-database/tests/test-index.js]: Github also doesn't retroactively apply the link : fb5e0b545

File: bamboo/bootstrap-tf-deployment.sh [hunk 6]
```diff
@@ -56,7 +56,31 @@ echo "Deploying Cumulus data-persistence module to $DEPLOYMENT"
...
+    region = \"$AWS_REGION\"
+    dynamodb_table = \"$TFSTATE_LOCK_TABLE\"
+  }
+}" >> ci_backend.tf
 
+# Initialize deployment
+../terraform init \
+  -input=false
+
+echo "Deploying the Cumulus data migration module to $DEPLOYMENT"
```
Comment (markdboyd) [inline bamboo/bootstrap-tf-deployment.sh:74]:
```suggestion
echo "Deploying the Cumulus database schema migration module to $DEPLOYMENT"
```

Nit

**Review by markdboyd — APPROVED** [review]

**Pull request merged**